### PR TITLE
Move some code out of headers

### PR DESCRIPTION
**Pull request by Luthaf**
This should improve compile time for rascal. I find locally a 10% decrease in compile time for a clean build with this PR, from 10 min to 9 min in release and from 12min to 11min in debug.

This mainly touches files in src/math and bindings. Other headers are template classes, so splitting the code between header and implementation will be harder. I can see two possibility: using extern template to declare explicitly instantiated templates, or using `.tpp` files with template code separated from the declarations.

---

Tasks before review:

- [x] benchmark overall performance to ensure the compiler do not miss optimization opportunities

TBD in a future PR:
- [ ] evaluate the use of `extern template` in the remaining of the code
- [ ] split test (they should not need header), and move tests utils to a separate place.

### CODE REVIEW

**Comment by max-veit** [comment]
Split & privatize? Who are you, Margaret Thatcher?

...anyway, yes, looks good, we've been meaning to do this for a while.

**Comment by Luthaf** [comment]
Here are the benchmarking results, using the existing hyp1f1 benchmarks

## Hyp1f1 function

### GCC

<details>

```
===============================  master  ======================================

Benchmark                     Time             CPU   Iterations nb_iterations
-----------------------------------------------------------------------------
bm_hyp1f1//0/0/0/0/0       44351 ns        44350 ns        15693          1000
bm_hyp1f1//0/1/0/0/0       46000 ns        45997 ns        15534          1000
bm_hyp1f1//0/2/0/0/0       44519 ns        44518 ns        15672          1000
bm_hyp1f1//0/3/0/0/0       44935 ns        44934 ns        15421          1000
bm_hyp1f1//0/0/1/0/0      482866 ns       482864 ns         1452           10k
bm_hyp1f1//0/1/1/0/0      483311 ns       483304 ns         1457           10k
bm_hyp1f1//0/2/1/0/0      479226 ns       479225 ns         1457           10k
bm_hyp1f1//0/3/1/0/0      478862 ns       478856 ns         1460           10k
bm_hyp1f1//0/0/2/0/0     4828981 ns      4828820 ns          145          100k
bm_hyp1f1//0/1/2/0/0     4836930 ns      4836911 ns          144          100k
bm_hyp1f1//0/2/2/0/0     4825121 ns      4825069 ns          145          100k
bm_hyp1f1//0/3/2/0/0     4848121 ns      4848053 ns          145          100k
bm_hyp1f1//0/0/3/0/0    48215176 ns     48214416 ns           14         1000k
bm_hyp1f1//0/1/3/0/0    48490732 ns     48490535 ns           15         1000k
bm_hyp1f1//0/2/3/0/0    48577528 ns     48577049 ns           14         1000k
bm_hyp1f1//0/3/3/0/0    48769563 ns     48769049 ns           14         1000k

===============================  this branch  =================================

Benchmark                      Time             CPU   Iterations nb_iterations
------------------------------------------------------------------------------
bm_hyp1f1//0/0/0/0/0       47872 ns        47865 ns        14602          1000
bm_hyp1f1//0/1/0/0/0       48492 ns        48484 ns        14672          1000
bm_hyp1f1//0/2/0/0/0       48468 ns        48467 ns        13397          1000
bm_hyp1f1//0/3/0/0/0       47549 ns        47547 ns        14687          1000
bm_hyp1f1//0/0/1/0/0      504797 ns       504550 ns         1371           10k
bm_hyp1f1//0/1/1/0/0      530718 ns       529111 ns         1340           10k
bm_hyp1f1//0/2/1/0/0      507791 ns       507789 ns         1387           10k
bm_hyp1f1//0/3/1/0/0      504969 ns       504962 ns         1399           10k
bm_hyp1f1//0/0/2/0/0     5218840 ns      5218507 ns          136          100k
bm_hyp1f1//0/1/2/0/0     5044321 ns      5043854 ns          136          100k
bm_hyp1f1//0/2/2/0/0     5046646 ns      5046627 ns          139          100k
bm_hyp1f1//0/3/2/0/0     5033305 ns      5033221 ns          100          100k
bm_hyp1f1//0/0/3/0/0    50345196 ns     50344541 ns           14         1000k
bm_hyp1f1//0/1/3/0/0    50396862 ns     50396287 ns           14         1000k
bm_hyp1f1//0/2/3/0/0    50215632 ns     50215023 ns           10         1000k
bm_hyp1f1//0/3/3/0/0    50290788 ns     50290157 ns           14         1000k

```

</details>

### CLANG

<details>

```
===============================  master  ======================================

Benchmark                      Time             CPU   Iterations nb_iterations
------------------------------------------------------------------------------
bm_hyp1f1//0/0/0/0/0       51962 ns        51888 ns        13572          1000
bm_hyp1f1//0/1/0/0/0       51224 ns        51149 ns        13281          1000
bm_hyp1f1//0/2/0/0/0       51549 ns        51473 ns        13505          1000
bm_hyp1f1//0/3/0/0/0       51478 ns        51398 ns        13276          1000
bm_hyp1f1//0/0/1/0/0      542995 ns       542222 ns         1274           10k
bm_hyp1f1//0/1/1/0/0      543537 ns       542770 ns         1294           10k
bm_hyp1f1//0/2/1/0/0      542747 ns       542098 ns         1273           10k
bm_hyp1f1//0/3/1/0/0      540750 ns       540046 ns         1275           10k
bm_hyp1f1//0/0/2/0/0     5477009 ns      5469616 ns          126          100k
bm_hyp1f1//0/1/2/0/0     5460726 ns      5452006 ns          127          100k
bm_hyp1f1//0/2/2/0/0     5463287 ns      5454565 ns          127          100k
bm_hyp1f1//0/3/2/0/0     5470486 ns      5461228 ns          127          100k
bm_hyp1f1//0/0/3/0/0    55556169 ns     55433082 ns           13         1000k
bm_hyp1f1//0/1/3/0/0    54946519 ns     54934762 ns           13         1000k
bm_hyp1f1//0/2/3/0/0    55120645 ns     55119927 ns           13         1000k
bm_hyp1f1//0/3/3/0/0    55363477 ns     55362815 ns           13         1000k

===============================  this branch  =================================


Benchmark                      Time             CPU   Iterations nb_iterations
------------------------------------------------------------------------------
bm_hyp1f1//0/0/0/0/0       52362 ns        52300 ns        13376          1000
bm_hyp1f1//0/1/0/0/0       52286 ns        52226 ns        13503          1000
bm_hyp1f1//0/2/0/0/0       52049 ns        51974 ns        13300          1000
bm_hyp1f1//0/3/0/0/0       51727 ns        51646 ns        13299          1000
bm_hyp1f1//0/0/1/0/0      554122 ns       553226 ns         1270           10k
bm_hyp1f1//0/1/1/0/0      551819 ns       551051 ns         1270           10k
bm_hyp1f1//0/2/1/0/0      551866 ns       551067 ns         1312           10k
bm_hyp1f1//0/3/1/0/0      550720 ns       550006 ns         1267           10k
bm_hyp1f1//0/0/2/0/0     5550173 ns      5541798 ns          125          100k
bm_hyp1f1//0/1/2/0/0     5593913 ns      5585082 ns          128          100k
bm_hyp1f1//0/2/2/0/0     5581626 ns      5574392 ns          125          100k
bm_hyp1f1//0/3/2/0/0     5566061 ns      5558755 ns          127          100k
bm_hyp1f1//0/0/3/0/0    56048604 ns     55963381 ns           13         1000k
bm_hyp1f1//0/1/3/0/0    55855635 ns     55782212 ns           13         1000k
bm_hyp1f1//0/2/3/0/0    55195539 ns     55115949 ns           13         1000k
bm_hyp1f1//0/3/3/0/0    56378353 ns     56300589 ns           12         1000k

```

</details>

**TL; DR**: master Hyp1f1 seems to be ~3-6% faster, but this is less than the difference between clang and gcc (gcc is ~10 - 15% faster)

## End to end benchmarks from Python side

<details>

```
===============================  master  ======================================

Benchmark on file tests/reference_data/small_molecule.json for optimization_args {'type': 'None'} with basis GTO
Timing SphericalExpansion
Finished in 580.41ns over 100 runs
Timing SphericalInvariants
Finished in 892.22ns over 100 runs

Benchmark on file tests/reference_data/small_molecule.json for optimization_args {'type': 'Spline', 'accuracy': 1e-08, 'range': (0, 4)} with basis GTO
Timing SphericalExpansion
Finished in 416.68ns over 100 runs
Timing SphericalInvariants
Finished in 781.02ns over 100 runs

Benchmark on file tests/reference_data/methane.json for optimization_args {'type': 'None'} with basis GTO
Timing SphericalExpansion
Finished in 280.84ns over 100 runs
Timing SphericalInvariants
Finished in 346.43ns over 100 runs

Benchmark on file tests/reference_data/methane.json for optimization_args {'type': 'Spline', 'accuracy': 1e-08, 'range': (0, 4)} with basis GTO
Timing SphericalExpansion
Finished in 255.73ns over 100 runs
Timing SphericalInvariants
Finished in 323.50ns over 100 runs

===============================  this branch  =================================

Benchmark on file tests/reference_data/small_molecule.json for optimization_args {'type': 'None'} with basis GTO
Timing SphericalExpansion
Finished in 571.12ns over 100 runs
Timing SphericalInvariants
Finished in 928.43ns over 100 runs

Benchmark on file tests/reference_data/small_molecule.json for optimization_args {'type': 'Spline', 'accuracy': 1e-08, 'range': (0, 4)} with basis GTO
Timing SphericalExpansion
Finished in 411.18ns over 100 runs
Timing SphericalInvariants
Finished in 785.11ns over 100 runs

Benchmark on file tests/reference_data/methane.json for optimization_args {'type': 'None'} with basis GTO
Timing SphericalExpansion
Finished in 283.43ns over 100 runs
Timing SphericalInvariants
Finished in 353.19ns over 100 runs

Benchmark on file tests/reference_data/methane.json for optimization_args {'type': 'Spline', 'accuracy': 1e-08, 'range': (0, 4)} with basis GTO
Timing SphericalExpansion
Finished in 254.49ns over 100 runs
Timing SphericalInvariants
Finished in 322.81ns over 100 runs

```

</details>

**TL; DR**: any difference between master and this branch is within the noise.

**Comment by max-veit** [comment]
Hang on, what are you trying to show with the benchmarks? Are we expecting any significant speed difference from moving code out of headers?

**Comment by Luthaf** [comment]
>  Are we expecting any significant speed difference from moving code out of headers?

No, but sometimes keeping code in header can improve performance. I am trying to show that moving code out of headers do not degrade performance.

File: src/rascal/math/spherical_harmonics.cc [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+ * along with this software; see the file LICENSE. If not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#include "rascal/math/spherical_harmonics.hh"
+
+#include <iostream>
+
+using namespace rascal::math;  // NOLINT
```
Comment (max-veit) [inline src/rascal/math/spherical_harmonics.cc]: Can we get rid of this `using namespace`? Or rather, if the alternative is to wrap the entire file in `namespace rascal::math`, then why does our style guide discourage `using namespace` statements for implementation files?

File: src/rascal/math/hyp1f1.hh [hunk 2]
```diff
@@ -30,487 +30,185 @@
...
+       public:
+        // Parameter used to fix the sum when evaluating the numerical
+        // derivatives.
+        int n_terms{100};
+
+        Hyp1f1Series(double a, double b, size_t mmax, double tolerance = 1e-14);
+
+        /**
+         * Computes G(a,b,z)
+         * @param z -> pow(alpha * r_ij, 2) / (alpha + beta)
```
Comment (max-veit) [inline src/rascal/math/hyp1f1.hh]: Might as well wrap these math statements in `@f`s, just to be consistent, in case this ever gets rendered.

File: src/rascal/math/utils.hh [hunk 3]
```diff
@@ -29,7 +29,7 @@
 #ifndef SRC_RASCAL_MATH_UTILS_HH_
 #define SRC_RASCAL_MATH_UTILS_HH_
 
-#include <Eigen/Dense>
+#include <Eigen/Core>
```
Comment (max-veit) [inline src/rascal/math/utils.hh]: Why? We never use anything else than `Dense`, do we?

File: tests/test_calculator.hh [hunk 4]
```diff
@@ -423,8 +423,8 @@ namespace rascal {
         "reference_data/small_molecule.json",
         "reference_data/methane.json"};
     // Simpler structures for debugging:
-    //"reference_data/diamond_2atom.json",
-    //"reference_data/SiC_moissanite.json",
+    // "reference_data/diamond_2atom.json",
+    // "reference_data/SiC_moissanite.json",
```
Comment (max-veit) [inline tests/test_calculator.hh]: Ugh, was this `cpplint` again? This extra space can make uncommenting a real PITA.

File: src/rascal/math/hyp1f1.hh [hunk 5]
```diff
@@ -30,487 +30,185 @@
...
-          coeff_derivative(0) = (a + 1) / (b + 1);
-          for (size_t i{1}; i < mmax; ++i) {
-            u = (a + i) / ((i + 1) * (b + i));
-            coeff(i) = coeff(i - 1) * u;
-            u = (a + i + 1) / ((i + 1) * (b + i + 1));
-            coeff_derivative(i) = coeff_derivative(i - 1) * u;
-          }
-        }
-      }
+    namespace internal {
```
Comment (max-veit) [inline src/rascal/math/hyp1f1.hh]: Why the move to `internal`? That namespace was only meant for things that don't need to be seen outside that file (or outside a group of two or three files). Furthermore, none of the other maths utilities are in this namespace, so this move is going to be confusing for everyone.

File: src/rascal/math/spherical_harmonics.cc (hunk 1 above)
Comment (Luthaf) [reply src/rascal/math/spherical_harmonics.cc]: >  then why does our style guide discourage using namespace statements for implementation files?

The usual rational is to forbid `using namespace xxx` in headers files, since this introduces namespace pollution in every file including the header. For implementation / .cc files, this is usually fine, since the `using namespace xxx` is scoped to this one and only file.

As far as the style guide goes, I have no idea.

File: tests/test_calculator.hh (hunk 4 above)
Comment (Luthaf) [reply tests/test_calculator.hh]: yes, cpplint.

You can configure your editor to comment / uncomment with spaces automatically.

File: src/rascal/math/hyp1f1.hh (hunk 5 above)
Comment (Luthaf) [reply src/rascal/math/hyp1f1.hh]: Because these classes are not used outside of this file. The only 'external' class here is `Hyp1f1SphericalExpansion`

Comment (Luthaf) [reply src/rascal/math/hyp1f1.hh]: >  Furthermore, none of the other maths utilities are in this namespace, so this move is going to be confusing for everyone.

For utilities only used inside `rascal::math`, I think they should be moved to `internal` too then. For utilities used outside of `rascal::math`, they should stay in there.

File: src/rascal/math/hyp1f1.hh (hunk 2 above)
Comment (Luthaf) [reply src/rascal/math/hyp1f1.hh]: Done!

File: src/rascal/math/spherical_harmonics.cc (hunk 1 above)
Comment (max-veit) [reply src/rascal/math/spherical_harmonics.cc]: Ok then, this usage seems sensible enough to me, so I’d suggest updating the style guide if it doesn’t already allow it (we might need to do this in the next group meeting). If possible, I’d also like to update the clang-format file to allow using-namespace declarations in implementation files; those `NOLINT` directives are super ugly.

File: src/rascal/math/utils.hh (hunk 3 above)
Comment (max-veit) [reply src/rascal/math/utils.hh]: Ah, I thought it was the other way around. So then the only file that really needs `Dense` is `calculator_spherical_expansion.hh` (since it uses an eigenvalue decomposition).

File: src/rascal/math/hyp1f1.hh (hunk 5 above)
Comment (max-veit) [reply src/rascal/math/hyp1f1.hh]: But the reason I said this was confusing is because these are high-level maths utilities, like calculating the hypergeometric function or modified spherical Bessel or whatever, that should be visible in the `rascal::math` namespace (which is already sequestered away from the main rascal namespace, so the rationale for having another `internal` to prevent namespace pollution is even weaker). The “main” classes for each file (like Hyp1f1 and its SphericalExpansion variant here) should be visible in `rascal::math`, if anything to aid discoverability in case someone wants to implement a new representation and would like to see what math functions we already implement.

It’s fine to put “internal” classes, like `Hyp1f1Series` and `Hyp1f1Asymptotic`, into `internal` because they’re not *meant* to be used outside the math utils. But the main classes, which are meant to be useful outside `rascal::math`, should stay as part of the math interface.

File: tests/test_calculator.hh (hunk 4 above)
Comment (max-veit) [reply tests/test_calculator.hh]: I know, but it’s still a pain. I’m used to thinking of comments with a space as “text comments” (explanatory text in plain English) and those without a space as temporarily disabled code.

File: src/rascal/math/spherical_harmonics.cc (hunk 1 above)
Comment (Luthaf) [reply src/rascal/math/spherical_harmonics.cc]: > those NOLINT directives are super ugly.

This is a limitation of cpplint. The rule is either enabled or disabled for all files in a folder, and I could not find a way to separate files by extension. We could have separated folders for .cc and .hh files, with separated CPPLINT.cfg, but I remember this was not the preferred solution when we discussed #189.

File: src/rascal/math/hyp1f1.hh (hunk 5 above)
Comment (Luthaf) [reply src/rascal/math/hyp1f1.hh]: > should be visible in rascal::math, if anything to aid discoverability in case someone wants to implement a new representation and would like to see what math functions we already implement.

I don't see how namespaces aid or prevent discoverability, but I agree anyway that `Hyp1f1` could also live in `rascal::math`. I mainly wanted to move `Hyp1f1Series` and `Hyp1f1Asymptotic` to `internal`, and `Hyp1f1` followed since it was not used by itself. I'll move it back.

File: src/rascal/math/utils.hh (hunk 3 above)
Comment (Luthaf) [reply src/rascal/math/utils.hh]: Might be, yes. I'd rather not change this in this PR and keep it for later

Comment (max-veit) [reply src/rascal/math/utils.hh]: Works for me.

File: src/rascal/math/spherical_harmonics.cc (hunk 1 above)
Comment (max-veit) [reply src/rascal/math/spherical_harmonics.cc]: Ugh, I knew it wasn’t going to be that simple. Let’s discuss it at the next group meeting, then.

File: src/rascal/math/hyp1f1.hh (hunk 5 above)
Comment (max-veit) [reply src/rascal/math/hyp1f1.hh]: As an example for discoverability: Let’s say we decide to make the API documentation more manageable by excluding anything within namespaces called `internal` (which would certainly be a reasonable suggestion). But this would then remove anything inside `rascal::math::internal` from the public documentation, perhaps leading a new developer to believe that we don’t have a dedicated Hyp1f1 class (just the specialization for spherical expansion). Of course, once they go into the code they’d find it, but we don’t want to make people search the source code for what is effectively API documentation.

Comment (Luthaf) [reply src/rascal/math/hyp1f1.hh]: updated

Comment (max-veit) [reply src/rascal/math/hyp1f1.hh]: Where? Looks like `Hyp1f1` is still in `namespace internal`.

Comment (Luthaf) [reply src/rascal/math/hyp1f1.hh]: Nope, see https://github.com/cosmo-epfl/librascal/pull/194/files#diff-18b2af0cf0528f03643f85559ceaaa0dR145 (GH getting confused by force-pushes again?)

Comment (max-veit) [reply src/rascal/math/hyp1f1.hh]: Ah, I was looking at an old local copy. Github isn't the only one getting confused by all this rewriting of history...
